File: node/test/base/limiter.test.js
/**
 * Copyright (c) 2015-2022, FineX, All Rights Reserved.
 * @author Thomas.Zhuang
 * @date 2022/05/18
 */

import { defaultLimiter, sleep } from '../../src/base/limiter';

export const limiterTest = 'limiterTest';

describe('limiter test', () => {
    test('max speed', async () => {
        let count = 0;
        for (let i = 0; i < 15; i++) {
            defaultLimiter.tryBeforeRun().then(() => {
                console.log(new Date());
                count++;
            });
        }
        await sleep(2100);// 15个满速情况需要2s即可(0,1,2s)
        console.log('total run:', count);
    });

    test('slow speed', async () => {
        for (let i = 0; i < 3; i++) {
            await sleep(1000);
            defaultLimiter.tryBeforeRun().then(() => {
                console.log(new Date());
            });
        }
    });
});

File: csharp/test/base/Limiter.Test.cs
/**
* Copyright (c) 2015-2022, FineX, All Rights Reserved.
* @author Thomas.Zhuang
* @date 2022/05/31
*/
using Src.Base;

namespace Test;
public class LimiterTest
{
    [Fact]
    public void TestTryBeforeRunMaxSpeed()
    {
        Thread.Sleep(1000);
        long start = new DateTimeOffset(DateTime.UtcNow).ToUnixTimeMilliseconds();
        List<ManualResetEvent> manualEvents = new List<ManualResetEvent>();
        for (int i = 0; i < 15; i++)
        {
            ManualResetEvent mre = new ManualResetEvent(false);
            manualEvents.Add(mre);
            ThreadPool.QueueUserWorkItem(Global.TryBeforeRun, mre);
        }
        WaitHandle.WaitAll(manualEvents.ToArray());
        long cost = new DateTimeOffset(DateTime.UtcNow).ToUnixTimeMilliseconds() - start;
        Console.WriteLine(cost);
    }

    // [Fact]
    // public void TestTryBeforeRunSlowSpeed()
    // {
    // }
}